monix-reactive/shared/src/main/scala/monix/reactive/internal/builders/Pad2Observable.scala

```diff
@@ -0,0 +1,191 @@
+ * limitations under the License.
+ */
+
+import monix.execution.cancelables.CompositeCancelable
+import monix.execution.{Cancelable, Ack}
+import monix.execution.Ack.{Stop, Continue}
+import monix.reactive.Observable
+import monix.reactive.observers.Subscriber
+import scala.concurrent.{Promise, Future}
+import scala.util.Success
+import scala.util.control.NonFatal
+
+private[reactive] final
+class Pad2Observable[A1,A2,+R]
+(obsA1: Observable[A1], obsA2: Observable[A2], paddingA1: ⇒ A1, paddingA2: ⇒ A2)
+(f: (A1,A2) ⇒ R)
+  extends Observable[R] { self ⇒
+
+  def unsafeSubscribeFn(out: Subscriber[R]): Cancelable = {
+    import out.scheduler
+
+    // MUST BE synchronized by `self`
+    var isDone = false
+    // MUST BE synchronized by `self`
+    var lastAck = Continue : Future[Ack]
+    // MUST BE synchronized by `self`
+    var elemA1 = null.asInstanceOf[A1]
+    // MUST BE synchronized by `self`
+    var hasElemA1 = false
+    // MUST BE synchronized by `self`
+    var elemA2 = null.asInstanceOf[A2]
+    // MUST BE synchronized by `self`
+    var hasElemA2 = false
+    // MUST BE synchronized by `self`
+    var continueP = Promise[Ack]()
+    // MUST BE synchronized by `self`
+    var completedCount = 0
+
+    // MUST BE synchronized by `self`
+    def rawOnNext(a1: A1, a2: A2): Future[Ack] =
+      if (isDone) Stop else {
+        var streamError = true
+        try {
+          val c = f(a1,a2)
+          streamError = false
+          out.onNext(c)
+        } catch {
+          case NonFatal(ex) if streamError =>
+            isDone = true
+            out.onError(ex)
+            Stop
+        } finally {
+          hasElemA1 = false
+          hasElemA2 = false
+        }
+      }
+
+    // MUST BE synchronized by `self`
+    def signalOnNext(a1: A1, a2: A2): Future[Ack] = {
+      lastAck = lastAck match {
+        case Continue => rawOnNext(a1,a2)
+        case Stop => Stop
+        case async =>
+          async.flatMap {
+            // async execution, we have to re-sync
+            case Continue => self.synchronized(rawOnNext(a1,a2))
+            case Stop => Stop
+          }
+      }
+
+      continueP.tryCompleteWith(lastAck)
+      continueP = Promise[Ack]()
+      lastAck
+    }
+
+    def signalOnError(ex: Throwable): Unit = self.synchronized {
+      if (!isDone) {
+        isDone = true
+        out.onError(ex)
+        lastAck = Stop
+      }
+    }
+
+    def signalOnComplete(hasElem: Boolean): Unit = {
+      @inline def rawOnComplete(): Unit =
+        if (!isDone) {
+          isDone = true
+          out.onComplete()
+        }
+
+      self.synchronized  {
+        val shouldComplete = !isDone && (
+          /*!hasElem &&*/ {
+            completedCount += 1
+            completedCount == 2
+          })
+        if (shouldComplete) {
+          lastAck match {
+            case Continue => rawOnComplete()
+            case Stop => () // do nothing
+            case async =>
+              async.onComplete {
```

Comment:
Here we do need to back-pressure, but instead of doing this `match`, just do `lastAck.syncOnContinue`.